packages/wrangler/src/r2/sql.ts

```diff
@@ -0,0 +1,227 @@
+
+		logger.log("Enabling R2 SQL for your account...");
+
+		try {
+			await fetchResult(config, `/accounts/${accountId}/dqe/enable`, {
+				method: "POST",
+				headers: {
+					"Content-Type": "application/json",
+				},
+				body: JSON.stringify({}),
+			});
+
+			logger.log(`✅ R2 SQL is enabled for your account
+
+Try sending a query with \`wrangler r2 sql query <warehouse name> <SQL query>\`
+`);
+		} catch (error) {
+			throw new UserError(
+				`Failed to enable R2 SQL: ${error instanceof Error ? error.message : String(error)}`
+			);
+		}
+	},
+});
+
+export const r2SqlDisableCommand = createCommand({
+	metadata: {
+		description:
+			"Disable sending SQL queries to R2 Data Catalogs in this account",
+		status: "open-beta",
+		owner: "Product: R2 SQL",
+	},
+	async handler(args, { config }) {
+		const accountId = await requireAuth(config);
+
+		logger.log("Disabling R2 SQL for your account...");
+
+		try {
+			await fetchResult(config, `/accounts/${accountId}/dqe/disable`, {
+				method: "POST",
+				headers: {
+					"Content-Type": "application/json",
+				},
+				body: JSON.stringify({}),
+			});
+
+			logger.log("✅ R2 SQL is disabled for your account");
+		} catch (error) {
+			throw new UserError(
+				`Failed to disable R2 SQL: ${error instanceof Error ? error.message : String(error)}`
+			);
+		}
+	},
+});
+
+export const r2SqlQueryCommand = createCommand({
+	metadata: {
+		description: "Execute SQL query against R2 Data Catalog",
+		status: "open-beta",
+		owner: "Product: R2",
+	},
+	positionalArgs: ["warehouse", "query"],
+	args: {
+		warehouse: {
+			describe: "R2 Data Catalog warehouse name",
+			type: "string",
+			demandOption: true,
+		},
+		query: {
+			describe: "The SQL query to execute",
+			type: "string",
+			demandOption: true,
+		},
+	},
+	async handler(args, { config: _config }) {
+		const { warehouse, query } = args;
+
+		const token = getR2SqlAPITokenFromEnv();
+		if (!token) {
+			// TODO: provide documentation link.
+			throw new UserError(
+				"CLOUDFLARE_R2_SQL_TOKEN environment variable is not set. " +
+					"Please set it to authenticate with the R2 SQL query service."
+			);
+		}
+
+		// TODO: add a spinner.
+		const splitIndex = warehouse.indexOf("_");
+		if (splitIndex === -1) {
+			throw new UserError("Warehouse name has invalid format");
+		}
+		const [accountId, bucketName] = [
+			warehouse.slice(0, splitIndex),
+			warehouse.slice(splitIndex + 1),
+		];
+
+		let responseStatus = null;
+		let text = null;
+		let duration = null;
+		try {
+			const start = Date.now();
+			const response = await fetch(
+				`https://api.dqe.cloudflarestorage.com/api/v1/accounts/${accountId}/dqe/query/${bucketName}`,
+				{
+					method: "POST",
+					headers: {
+						Authorization: `Bearer ${token}`,
+						"Content-Type": "application/json",
+					},
+					body: JSON.stringify({
+						warehouse,
+						query,
+					}),
+				}
+			);
+			responseStatus = response.status;
+			text = await response.text();
```

Comment:
Given this is being parsed to JSON, this could just be `response.json()`?

Reply:
Two reasons:
- Mention text in the error message if it's not a valid json. Calling `response.json()` and then `response.text()` in case of exception is not safe because we might have already read all the text from the connection in `response.json()`
- To separate "connection errors" and "invalid json errors" in exceptions we throw. This might not be that useful

Reply:
I see, makes sense. This mostly looks okay, but we already have a utility for making fetch requests in Wrangler (`fetchResult()`) that handles all this + debug logging etc... Can that be re-used at all? Perhaps with an additional parameter for overriding the base URL?

Reply:
I don't think this can be done easily. `fetchResult` requires the user to login and passes the API token in the headers. We technically don't need the user to login and it can be confusing to show login prompt and then still ask for the token. `fetchResult` also adds a bunch of headers we do not need. It can still be done, but I think it will complicated already quite involved `fetchResult` and won't save us much effort for this specific command.